core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```diff
@@ -1450,12 +1457,89 @@ class GroupCoordinator(val brokerId: Int,
             group.maybeInvokeJoinCallback(member, joinResult)
             completeAndScheduleNextHeartbeatExpiration(group, member)
             member.isNew = false
+
+            group.addPendingSyncMember(member.memberId)
           }
+
+          schedulePendingSync(group)
         }
       }
     }
   }
 
+  private def maybeRemovePendingSyncMember(
+    group: GroupMetadata,
+    memberId: String
+  ): Unit = {
+    group.removePendingSyncMember(memberId)
+    maybeCompleteSyncExpiration(group)
+  }
+
+  private def removeSyncExpiration(
+    group: GroupMetadata
+  ): Unit = {
+    group.clearPendingSyncMembers()
+    maybeCompleteSyncExpiration(group)
+  }
+
+  private def maybeCompleteSyncExpiration(
+    group: GroupMetadata
+  ): Unit = {
+    val groupKey = GroupKey(group.groupId)
+    syncPurgatory.checkAndComplete(groupKey)
+  }
+
+  private def schedulePendingSync(
+    group: GroupMetadata
+  ): Unit = {
+    val delayedSync = new DelayedSync(this, group, group.rebalanceTimeoutMs)
```

Comment:
I was considering if we should start the timer fresh after the group is joined. An alternative is to let the rebalance timeout cover the total time from when the rebalance is first triggered. On the other hand, if we are trying to allow for a full max.poll.interval.ms in between Join and Sync, then maybe we need to reset it.

Reply:
That's a good point. I thought about it as well. Given that the goal of this PR is to protect us against misbehaving or buggy clients, I think that it is OK to allow for a full max.poll.interval.ms in between Join and Sync. Practically, I have found this approach a bit harder to test as we end up with both a DelayedJoin and a DelayedSync in parallel and they both rely on the same rebalance timeout. It is a little easier to reason about them when the are disjoint. That being said, I don't feel strong either ways.

Reply:
Resetting the timer sounds fine to me.